### PR TITLE
feat(react): FiberNodeButton extensibility APIs for tabs/actions/i18n

### LINKED ISSUE #144: FiberNodeButton 面向开发者二次开发灵活性评估与改进提案
## 背景
当前 `FiberNodeButton` 的连接与支付主流程已经比较完善，开箱即用体验很好。

## 现状结论
封装质量较好，但在 **面板内容与功能的二次开发灵活性** 上属于中等偏低。

### 现有灵活点
- 支持传入外部 `fiber` 实例，便于状态共享
- 支持 `externalFunding` resolver，外部钱包流程可接
- 暴露 `onConnect` / `onDisconnect` / `onError` / `onLog` 回调
- 提供 `renderConnectorSection` 插槽用于连接区扩展

### 主要限制
- Tab 结构固定（Workbench / Channels / Diagnostics），无法按业务自由增删改
- 各 Tab 内字段与动作以内建逻辑为主，局部替换成本高
- 文案与布局基本内聚在组件内部，消费者难以定制

## 建议方向
采用“双层能力”策略：
1. 保留现有开箱即用版 `FiberNodeButton`（兼容当前用户）
2. 增加可编排扩展能力，支持业务定制而不必 fork

## 最小可行改进（MVP）
- `tabs` 配置：支持重排 / 隐藏 / 新增自定义 tab
- `renderTabContent(tabId, context)`：允许覆盖单个 tab 内容
- `actions` 覆盖：允许替换默认 action（如 open channel / pay / close channel）
- 文案配置入口（i18n-ready）

## 验收建议
- 不破坏现有 `FiberNodeButton` props
- 默认行为与 UI 与当前版本一致
- demo 增加“自定义 tab 内容”示例

### PR DESCRIPTION
**Pull request by RetricSu**
## Summary
- add extensibility APIs to FiberNodeButton: tabs, renderTabContent, renderAction, and t i18n function
- refactor panel rendering to support tab reorder/hide/custom tabs with active-tab fallback
- add action rendering abstraction so open/pay/create/close/force-close can be overridden without forking
- add react-quick-card Custom Tab Mode demo and README customization docs
- add extensibility test suite for default compatibility and new extension paths

## Validation
- pnpm -w lint
- cd packages/react && pnpm test -- --run && pnpm build
- full pre-commit checks passed (build/typecheck/test across workspace)

Closes #144

### CODE REVIEW

**Review by humble-little-bear — COMMENTED** [review]
## 总体评价

这个 PR 做得非常扎实。在不破坏默认行为的前提下，为 `FiberNodeButton` 引入了清晰的扩展点（tabs / renderTabContent / renderAction / i18n），并且测试、文档、Demo 都同步跟进了。代码结构和类型设计都很合理，作为 SDK 的扩展性改造，这是一个高质量的实现。

---

## 做得好的地方

1. **Additive API 设计**：所有扩展都是 opt-in，默认行为完全不变。`resolveTabs` 的"用户配置优先 + 未提及默认项自动追加"策略很符合直觉。
2. **类型系统完善**：`FiberNodeButtonActionDefaultProps` 用 discriminated union 区分不同 action 的 props，`satisfies` 在调用处保证了类型安全，同时保留了推断灵活性。
3. **i18n 实现轻量**：`(key, fallback, vars?)` 的签名简单实用，默认实现直接返回带变量替换的 fallback，零成本接入。
4. **Demo 和文档齐全**：`react-quick-card` 的 Custom Tab Mode 演示了中文本地化、自定义 Tab、自定义 Action 的完整链路；README 的示例代码可以直接复制使用。
5. **测试覆盖到位**：默认兼容性 + 扩展路径都有了测试，验证了 tab 重排、隐藏、自定义内容、自定义 action、i18n 这几个核心场景。
6. **A11y 保持**：tablist / tab / tabpanel 的 role 和 aria 属性在重构后仍然完整。

---

## 建议与顾虑

### 1. `tabContext` 的引用稳定性（性能）

`tabActions` 的依赖是 `[state]`，而 `state` 是 `useFiberNodeButtonPanelState` 返回的整个对象，每次 render 都会换新引用。这导致 `tabActions` → `tabContext` → `resolvedTabs` 链条上的 memo 都失效，自定义 tab 的 `render` 和 `renderTabContent` 每次都会收到新的 `context` 对象。

**建议**：如果未来这个组件出现在高频刷新的场景里，可以考虑让 `useFiberNodeButtonPanelState` 返回一个 `useMemo` 包裹的 stable state 对象，或者让 `tabActions` 直接依赖 `state` 里真正用到的几个函数（`openChannel`, `createInvoice` 等）。不过这是一个 pre-existing 问题，不是这个 PR 引入的，可以后续优化。

### 2. `renderAction` 的 context 未暴露 `state`

当前 `FiberNodeButtonRenderActionContext` 只有 `id`, `defaultProps`, `fiber`, `t`。如果用户想做一个自定义的 "Open Channel" 按钮，在按钮上显示当前输入的 peerPubkey 或 fundingAmount，他无法从 context 里拿到这些数据，只能依赖外部闭包。

**建议**：考虑把 `state`（或精简后的 `state` 片段）也放进 `FiberNodeButtonRenderActionContext`。当然，如果设计意图就是让 `renderAction` 只管按钮 UI/行为、不管周边数据展示，那现在的设计是合理的。

### 3. 全隐藏默认 tab 的成本

目前默认 tab 是"自动追加"策略。如果用户只想保留一个完全自定义的 Tab 面板，需要把 `workbench` / `channels` / `diagnostics` 逐个设 `hidden: true`。

**建议**：这不是严重问题，但如果后续收到反馈，可以考虑加一个 `tabs` 的语义：如果用户传了 `tabs` 且其中不包含任何默认 tab，则视为完全接管，不再自动追加默认项。当然这会改变现有语义，需要谨慎。

### 4. Edge case：所有 tab 都被隐藏

当 `resolvedTabs.length === 0` 时，effect 里会 `return`，不会切换 activeTab。但此时 `tabContent` 仍然会尝试根据 `activeTab` 渲染，且 tablist 为空。虽然现实中不太可能有人这么用，但可以考虑在 `resolvedTabs.length === 0` 时给一个空面板或占位提示。

### 5. `null` vs `undefined` 在 `renderTabContent` 中的语义

当前逻辑是 `overriddenTabContent === undefined` 才走默认渲染。如果用户返回 `null`，会渲染空白。这个行为是对的（`null` = 明确不渲染），但可以考虑在文档里提一句，避免用户困惑。

---

## 小疑问

- `packages/react/src/fiber-node-button/index.tsx` 里 `export type { ... } from './types.js'` 的写法在包入口 `src/index.ts` 里又重复 export 了一次。这是有意为之的 double-export 吗？如果是，确保不会导致用户侧看到重复的类型声明提示。

---

## 结论

这个 PR 的设计思路清晰，实现干净，文档和测试都到位。上面的建议都是优化项或未来可改进的方向，不影响合并。**LGTM** 👍

File: packages/react/src/fiber-node-button/panel.tsx [hunk 1]
```diff
@@ -26,12 +118,100 @@ export function FiberNodeButtonPanel(props: FiberNodeButtonPanelProps) {
...
+      },
+      closeChannel: async (channelId: string) => {
+        await state.closeChannel(channelId, false);
+      },
+      forceCloseChannel: async (channelId: string) => {
+        await state.closeChannel(channelId, true);
+      },
+    }),
+    [state],
+  );
```
[the remark this answers is not in the document]
Comment (RetricSu) [reply packages/react/src/fiber-node-button/panel.tsx:159]: Good catch. Updated in commit 9ec2d2a: tabActions no longer depends on the full state object. It now depends only on stable action callbacks (openChannel/createInvoice/submitPayment/closeChannel), which removes this avoidable memo invalidation path.

File: packages/react/src/fiber-node-button/panel.tsx [hunk 2]
```diff
@@ -26,12 +118,100 @@ export function FiberNodeButtonPanel(props: FiberNodeButtonPanelProps) {
...
+    }),
+    [externalFunding?.enabled, fiber, state, t, tabActions],
+  );
+
+  const resolvedTabs = useMemo(() => resolveTabs(tabs, t), [t, tabs]);
+
+  const tabListStyle = useMemo(
+    () => ({
+      ...styles.tabList,
+      gridTemplateColumns: `repeat(${Math.max(1, resolvedTabs.length)}, minmax(0, 1fr))`,
```
[the remark this answers is not in the document]
Comment (RetricSu) [reply packages/react/src/fiber-node-button/panel.tsx]: Implemented in commit 9ec2d2a. The tab list now uses `minmax(112px, 1fr)` and `overflowX: auto` in the computed tab-list style, so many custom tabs can scroll horizontally instead of shrinking into unreadable widths.

File: packages/react/src/fiber-node-button/panel.tsx [hunk 3]
```diff
@@ -26,12 +118,100 @@ export function FiberNodeButtonPanel(props: FiberNodeButtonPanelProps) {
...
   }, [forceCloseConfirmOpen]);
 
+  useEffect(() => {
+    if (resolvedTabs.length === 0) {
+      return;
+    }
+    if (!resolvedTabs.some((tab) => tab.id === activeTab)) {
+      switchTab(resolvedTabs[0].id);
+    }
+  }, [activeTab, resolvedTabs, switchTab]);
```
[the remark this answers is not in the document]
Comment (RetricSu) [reply packages/react/src/fiber-node-button/panel.tsx:245]: Addressed in commit 9ec2d2a. I introduced `effectiveActiveTab` (with fallback to first visible tab) and switched rendering/ARIA bindings to use the effective tab instead of raw state. This avoids transient invalid `aria-labelledby`/tabpanel references before state sync effect runs. Also added an empty-tab notice when all tabs are hidden.

File: packages/react/src/fiber-node-button/panel.tsx [hunk 4]
```diff
@@ -26,12 +118,100 @@ export function FiberNodeButtonPanel(props: FiberNodeButtonPanelProps) {
...
+      tabContent = (
+        <ChannelsTab state={state} onLog={onLog} renderAction={renderAction} t={t} fiber={fiber} />
+      );
+    } else if (activeTab === 'diagnostics') {
+      tabContent = <DiagnosticsTab state={state} t={t} />;
+    } else if (selectedResolvedTab?.render) {
+      tabContent = selectedResolvedTab.render(tabContext);
+    } else {
+      tabContent = (
+        <div style={styles.notice}>
```
[the remark this answers is not in the document]
Comment (RetricSu) [reply packages/react/src/fiber-node-button/panel.tsx]: Implemented option (a) in commit 9ec2d2a: `tabs[i].render` now takes precedence over built-in tab bodies (after `renderTabContent`), including built-in ids. I also added regression tests for this behavior and clarified precedence + null/undefined semantics in the README.

**Comment by RetricSu** [comment]
@humble-little-bear 感谢详细 review，我逐点看过并补了一个直接可落地的改进：\n\n- 已在 commit ec843fa 中把  context 补充了 ，现在自定义 action 可以直接读取 panel 当前状态（例如 peerPubkey/fundingAmount 等），不必依赖外部闭包。\n- 这次也同步更新了文档与测试（新增断言 renderAction 可拿到 state）。\n\n关于你提到的其他点：\n-  的  vs  语义与渲染优先级已在 README 明确。\n- “所有 tab 都 hidden”的空态已在面板里处理。\n-  引用稳定性（state 整体引用）我认同是值得优化的点，计划后续在  做更细粒度 memo/stable-state 改造，避免本 PR 再扩大范围。\n\n再次感谢，建议很有帮助 🙏

**Comment by RetricSu** [comment]
@humble-little-bear 感谢详细 review，我逐点看过并补了一个可直接落地的改进。

- 已在 commit ec843fa 中把 renderAction context 补充了 state，现在自定义 action 可以直接读取 panel 当前状态（例如 peerPubkey/fundingAmount 等），不必依赖外部闭包。
- 这次也同步更新了文档与测试（新增断言 renderAction 可拿到 state）。

关于你提到的其他点：
- renderTabContent 的 null vs undefined 语义与渲染优先级已在 README 明确。
- 所有 tab 都 hidden 的空态已在面板里处理。
- tabContext 引用稳定性（state 整体引用）我认同是值得优化的点，计划后续在 useFiberNodeButtonPanelState 做更细粒度 memo/stable-state 改造，避免本 PR 再扩大范围。

再次感谢，建议很有帮助。
